asf_vahp_visitor_evaluation/pipeline/load_data.py

```diff
@@ -0,0 +1,312 @@
+            # If multiple complete, use first by time submitted
+            drop.extend(
+                candidates.loc[candidates["Status"] == "Complete", ["Date Submitted"]]
+                .sort_values("Date Submitted")
+                .index[1:]
+                .to_list()
+            )
+        elif len(candidates.loc[candidates["Status"] == "Complete", :]) == 1:
+            # if only 1 complete, use that one
+            drop.extend(
+                candidates.loc[candidates["Status"] != "Complete", :].index.to_list()
+            )
+        else:
+            # Otherwise use the earliest partial
+            drop.extend(candidates["Date Submitted"].sort_values().index[1:].to_list())
+    # Drop duplicate indices
+    df = df.drop(index=drop).reset_index(drop=True)
+    return df
+
+
+def make_analysis_ready(
+    source: str,
+    usecols: Iterable,
+    dtype: dict,
+    date_cols: list,
+    categories: dict,
+    binaries: list,
+    wave_id: str,
+    visitor_lookup: str,
+) -> pandas.DataFrame:
+    """Generic function for loading VAHP survey data.
+
+    Fixes issues with unicode characters and strips whitespace.
+    Converts data to datetime, categorical and binary where relevant.
+    Anonymises data for analysis by removing email identifier and adding
+    numerical id for linkage. Removes non-consenting responses.
+
+    Args:
+        source: Filepath to survey data to be processed.
+        usecols: Iterable of column names to load from source.
+        dtype: Schema dictionary for columns at read.
+        data_cols: List of columns to interpret as datetime objects.
+        categories: Dictionary of categorical columns (keys) with a tuple
+          of categories and order (e.g. {"col": (['cat1', 'cat2'], False),})
+        binaries: List of columns to be convert to boolean.
+        wave_id: String to indicate wave of data loaded.
+        visitor_lookup: Filepath to lookup file.
+
+    Returns:
+        A pandas DataFrame that is the processed analysis-ready version
+        of the input raw survey data.
+    """
+    # 1. Load raw data
+    data = pandas.read_csv(
+        source,
+        usecols=usecols,
+        dtype=dtype,
+    )
+    # 2. Fix column names
+    data.columns = _fix_column_names(list(data.columns))
+    # 3. Fix responses
+    data.loc[:, data.dtypes == "object"] = data.loc[:, data.dtypes == "object"].map(
+        _fix_cell
+    )
+    # 4. Convert dates
+    data = _parse_dates(data, date_cols)
+    # 5. Convert categories
+    data = _parse_categories(data, categories)
+    # 6. Convert binaries
+    data = _parse_binaries(data, binaries)
+    # 7. Add wave identifier at start of table
+    col = pandas.Series(data=numpy.repeat(wave_id, repeats=len(data)), name="wave")
+    data.insert(3, col.name, col)
+    # 8. Remove non-consenting responses
+    data = data.loc[
+        data[data.columns[data.columns.str.startswith("I agree to participate")][0]]
+        == "Yes",
+        :,
+    ]
```

Comment:
Suggestion for conciseness:

```suggestion
    data = data[data[previsitq.q0e] == "Yes"]
```

Reply:
Yes, good point, have revised.